src/ucp/core/ucp_am.c

```diff
@@ -0,0 +1,699 @@
+                         size_t am_length, ucp_ep_h reply_ep) {
+    uint16_t am_id;
+    ucs_status_t status;
+
+    memcpy(UCS_PTR_BYTE_OFFSET(unfinished->all_data + 1, long_hdr->offset),
+           long_hdr + 1, am_length - sizeof(*long_hdr));
+    unfinished->left -= am_length - sizeof(*long_hdr);
+    if (unfinished->left == 0) {
+        am_id = long_hdr->am_id;
+        status = worker->am_cbs[am_id].cb(worker->am_cbs[am_id].context,
+                                          unfinished->all_data + 1,
+                                          long_hdr->total_size,
+                                          reply_ep,
+                                          UCP_CB_PARAM_FLAG_DATA);
+
+        if (status != UCS_INPROGRESS) {
+            ucs_free(unfinished->all_data);
+        }
+
+        ucs_list_del(&unfinished->unfinished);
+        ucs_free(unfinished);
+    }
+    
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_am_long_handler_common(void *am_arg, void *am_data, size_t am_length,
+                           unsigned am_flags, ucp_ep_h reply_ep)
+{
+    ucp_worker_h worker = (ucp_worker_h) am_arg;
+    ucp_am_long_hdr_t *long_hdr = (ucp_am_long_hdr_t *) am_data;
+    ucp_ep_h ep = ucp_worker_get_ep_by_ptr(worker, long_hdr->ep);
+    ucp_ep_ext_proto_t *ep_ext = ucp_ep_ext_proto(ep);
+    ucp_recv_desc_t *all_data;
+    size_t left;
+    ucp_am_unfinished_t *unfinished;
+
+    if (ucs_unlikely((long_hdr->am_id >= worker->am_cb_array_len) ||
+		             worker->am_cbs[long_hdr->am_id].cb == NULL)) {
+        ucs_warn("UCP Active Message was received with id : %u, but there" 
+                 "is no registered callback for that id", long_hdr->am_id);
+        goto out;;
+    }
+
+    /* if there are multiple messages,
+     * we first check to see if any of the other messages
+     * have arrived. If any messages have arrived,
+     * we copy ourselves into the buffer and leave
+     */
+    unfinished = ucp_am_find_unfinished(worker, ep, ep_ext, long_hdr, am_length);
+    
+    if (unfinished) {
+        return ucp_am_handle_unfinished(worker, unfinished, 
+                                        long_hdr, am_length,
+                                        reply_ep);
+    }
+    
+    /* If I am first, I make the buffer for everyone to go into,
+     * copy myself in, and put myself on the list so people can find me
+     */
+    all_data = ucs_malloc(long_hdr->total_size
+                          + sizeof(ucp_recv_desc_t),
+                          "ucp recv desc for long AM");
+
+    all_data->flags = UCP_RECV_DESC_FLAG_MALLOC;
+
+    left = long_hdr->total_size - (am_length -
+                                   sizeof(ucp_am_long_hdr_t));
+    
+    memcpy(UCS_PTR_BYTE_OFFSET(all_data + 1, long_hdr->offset),
+           long_hdr + 1, am_length - sizeof(ucp_am_long_hdr_t));
+    
+    /* Can't use a desc for this because of the buffer */
+    unfinished              = ucs_malloc(sizeof(ucp_am_unfinished_t),
```

Comment:
why do we need a list of unfinished, if no out-of-order is supported? we expect to fully complete one AM before receiving the next one..
and with out-of-order, better use hash table anyway instead of linear search...

Reply:
It is designed to support out of order. Are ucp_eps guaranteed to be a one-to-one? tag_recv handled this similarly.

Reply:
For our use case we don't really care. IMHO, in a long term it is better to go with out-of-order. Users alway can use fence to order requests on EP.

Reply:
let's add TODO to convert this list to a hash for O(1) lookup